accepted/wasmtime-debugging.md

```diff
@@ -0,0 +1,493 @@
+This should allow us to have best-effort core dumps from cranelift without
+overly pessimizing optimization.
+
+While it will be possible to generate core dumps from code compiled with
+Cranelift, we do accept that it won’t have the same level of detail available to
+Winch. As the wasm core dump format is defined in terms of the wasm execution
+model and is independent of the strategy used to execute that wasm, we will be
+able to use core dumps generated by code produced by either Cranelift or Winch
+when debugging. Being able to use a core dump generated by code compiled with
+Cranelift when debugging with Winch is extremely powerful, as it means that we
+don’t need to sacrifice the performance of generated code to enable debugging
+with a core dump. Additionally, as Winch will have more information available
+when debugging, the core dump produced by code compiled with either backend will
+be sufficient to get the debugger into a state where we have high-fidelity
+information available.
+
+## Debug Adapter Protocol
+
+We plan to add initial support for the [Debug Adapter Protocol
+(DAP)](https://microsoft.github.io/debug-adapter-protocol/) into wasmtime. This
+will involve adding an execution mode like `wasmtime --dap` or `wasmtime
+--debug`, which will expose a debug adapter. The idea here is to allow DAP
+clients to consume core dump information from Wasmtime, enabling users to start
+doing post-mortem debugging, and acting as a forcing function for us to
+implement and verify basic protocol support before moving onto live debugging
+use-cases.
+
+Why the debug adapter protocol? [Like the Language Server Protocol
+(LSP)](https://microsoft.github.io/language-server-protocol/) the Debug Adapter
+Protocol allows for a client-agnostic implementation of the debugger.
+Traditionally tools like gdb or lldb provide their own front-end for interacting
+with programs that are being debugged, but this requires changing out of a
+user’s more familiar development environment when switching to a debugging
+workflow. The Debug Adapter Protocol provides a separation that allows for
+supporting editors (like
+[VSCode](https://code.visualstudio.com/api/extension-guides/debugger-extension),
+[Emacs](https://www.emacswiki.org/emacs/DebugAdapterProtocol), and
+[neovim](https://neovim.io/doc/user/lsp.html)) to directly implement setting
+breakpoints and interacting with programs that are being debugged, instead of
+requiring the use of a different tool. Targeting the Debug Adapter Protocol will
+allow Wasmtime developers to focus on the implementation of debugger primitives,
+rather than the user experience of debugging programs. In addition to the
+features we outline in the [Live Debugging](#live-debugging) section below, the
+DAP supports debugging debuggees that cannot step or set breakpoints (such as
+core dumps) and debuggees that can reverse step (such as time-traveling while
+replaying a recorded trace). Many other debugging protocols do not have these
```

Comment:
Both the gdbserver and gdb-mi interfaces support this. Rr makes use of it and gdb itself has a very basic reverse stepping implementation too.

Reply:
I'm happy to remove this sentence, though I would again point to @fitzgen's comments about ease of initial implementation for why we shouldn't try to use gdb in our initial implementation.